helix-core/src/main/java/org/apache/helix/task/TaskDriver.java

```diff
@@ -585,36 +585,195 @@ public void addTask(String workflowName, String jobName, TaskConfig taskConfig,
 
     long endTime = System.currentTimeMillis() + timeoutMs;
 
-    validateAddTaskConfigs(workflowName, jobName, taskConfig);
+    validateConfigsForTaskModifications(workflowName, jobName, taskConfig);
 
     String nameSpaceJobName = TaskUtil.getNamespacedJobName(workflowName, jobName);
+    JobConfig jobConfig = TaskUtil.getJobConfig(_accessor, nameSpaceJobName);
+    for (String taskEntry : jobConfig.getMapConfigs().keySet()) {
+      if (taskEntry.equals(taskConfig.getId())) {
+        throw new HelixException(
+            "Task cannot be added because another task with the same ID already exists!");
+      }
+    }
+
     WorkflowContext workflowContext = getWorkflowContext(workflowName);
     JobContext jobContext = getJobContext(nameSpaceJobName);
+    // If workflow context or job context is null. It means job has not been started. Hence task can
+    // be added to the job
+    if (workflowContext != null && jobContext != null) {
+      TaskState jobState = workflowContext.getJobState(nameSpaceJobName);
+      if (jobState != null && ILLEGAL_JOB_STATES_FOR_TASK_MODIFICATION.contains(jobState)) {
+        throw new HelixException("Job " + nameSpaceJobName
+            + " is in illegal state for task addition. Job State is " + jobState);
+      }
+    }
+
+    DataUpdater<ZNRecord> updater = currentData -> {
+      if (currentData != null) {
+        currentData.setMapField(taskConfig.getId(), taskConfig.getConfigMap());
+      } else {
+        LOG.error("JobConfig DataUpdater: Fails to update JobConfig. CurrentData is null.");
+      }
+      return currentData;
+    };
+
+    updateTaskInJobConfig(workflowName, jobName, updater);
+
+    workflowContext =
+        _accessor.getProperty(_accessor.keyBuilder().workflowContextZNode(workflowName));
+    jobContext =
+        _accessor.getProperty(_accessor.keyBuilder().jobContextZNode(workflowName, jobName));
+
     if (workflowContext == null || jobContext == null) {
-      // Workflow context or job context is null. It means job has not been started. Hence task can
-      // be added to the job
-      addTaskToJobConfig(workflowName, jobName, taskConfig, endTime);
       return;
     }
 
-    TaskState jobState = workflowContext.getJobState(nameSpaceJobName);
+    String taskID = taskConfig.getId();
+    while (System.currentTimeMillis() <= endTime) {
+      jobContext =
+          _accessor.getProperty(_accessor.keyBuilder().jobContextZNode(workflowName, jobName));
+      workflowContext =
+          _accessor.getProperty(_accessor.keyBuilder().workflowContextZNode(workflowName));
+      if (jobContext.getTaskIdPartitionMap().containsKey(taskID)
+          && workflowContext.getJobState(nameSpaceJobName) == TaskState.IN_PROGRESS) {
```

Comment:
Please make sure user understand it. If they have the delay scheduling between tasks or if there are dependencies, this add may fail. Also we do not do "rollback" for added tasks.

Reply:
Added some java docs note. Thanks.